### PR TITLE
Replace debs with tar balls

### PR DESCRIPTION
**Pull request by conr**

### CODE REVIEW

File: site.yml [hunk 1]
```diff
@@ -1,15 +1,15 @@
 ---
 - hosts: all
   tags: [base]
-  roles: 
-    - base
+  roles:
+    - { role: base, sudo: yes }
```
Comment (brian-brazil) [inline site.yml]: The root handling belongs in the ansible.cfg

File: roles/prometheus/templates/etc/systemd/system/node_exporter.service.j2 [hunk 2]
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Node_exporter
+After=network.target
+
+[Service]
+Type=simple
+PIDFile={{ node_exporter_pid_path }}
+ExecReload=/bin/kill -HUP $MAINPID
```
Comment (brian-brazil) [inline roles/prometheus/templates/etc/systemd/system/node_exporter.service.j2]: The node exporter doesn't do reloads

File: roles/prometheus/templates/etc/systemd/system/alertmanager.service.j2 [hunk 3]
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Alertmanager
+After=network.target
+
+[Service]
```
Comment (brian-brazil) [inline roles/prometheus/templates/etc/systemd/system/alertmanager.service.j2]: There should be a Restart=always here

File: roles/base/handlers/main.yml [hunk 4]
```diff
@@ -1,3 +1,3 @@
 ---
-- name: restart sshd
+- name: Restart sshd
```
Comment (brian-brazil) [inline roles/base/handlers/main.yml]: Keep the handlers lowercase

File: roles/base/tasks/main.yml [hunk 5]
```diff
@@ -1,22 +1,21 @@
...
-    content: |
-      deb http://deb.robustperception.io/ precise nightly
-
-- name: Add prometheus apt key
-  notify: update apt
-  apt_key: id=41EFC99D url=https://s3-eu-west-1.amazonaws.com/deb.robustperception.io/41EFC99D.gpg
+- name: Install required packages
+  apt:
+    name: "{{ item }}"
+    state: present
```
Comment (brian-brazil) [inline roles/base/tasks/main.yml]: present is the default, no need to mention it

File: roles/prometheus/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+---
+prometheus_version: 1.8.0
+prometheus_gomaxprocs: "{{ ansible_processor_vcpus|default(ansible_processor_count) }}"
```
Comment (brian-brazil) [inline roles/prometheus/defaults/main.yml]: this isn't needed

File: roles/prometheus/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,49 @@
+---
+prometheus_version: 1.8.0
```
Comment (brian-brazil) [inline roles/prometheus/defaults/main.yml]: Use a 2.0

File: roles/prometheus/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+prometheus_config_dir: /etc/prometheus
+prometheus_console_libs_dir: /etc/prometheus/console_libraries
+prometheus_console_dir: /etc/prometheus/consoles
+prometheus_log_dir: /var/log/prometheus
+prometheus_pid_path: /var/run/prometheus.pid
+prometheus_db_dir: /var/lib/prometheus
+
+prometheus_web_listen_address: ":9090"
+prometheus_web_external_url: 'http://localhost:9090/'
```
Comment (brian-brazil) [inline roles/prometheus/defaults/main.yml]: This will be demo.robustperception.io

File: roles/prometheus/defaults/main.yml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+prometheus_alertmanager_url: 'http://localhost:9093'
+
+prometheus_config_flags:
```
Comment (brian-brazil) [inline roles/prometheus/defaults/main.yml]: It's easier to hardcode all this in the unit file, we don't need these configurable

File: roles/prometheus/files/node.html [hunk 10]
```diff
@@ -0,0 +1,34 @@
+{{ template "head" . }}
```
Comment (brian-brazil) [inline roles/prometheus/files/node.html]: This comes from Prometheus, no need to duplicate it

File: roles/prometheus/handlers/main.yml [hunk 11]
```diff
@@ -1,7 +1,6 @@
 ---
-- name: reload prometheus
```
Comment (brian-brazil) [inline roles/prometheus/handlers/main.yml]: We should keep the reload

File: roles/prometheus/tasks/main.yml [hunk 12]
```diff
@@ -1,53 +1,113 @@
...
+  file: >
+    state=directory
+    path={{ item }}
+    owner=root
+    group=root
+  with_items:
+    - "{{ prometheus_root_dir }}"
+    - "{{ prometheus_dist_dir }}"
+    - "{{ prometheus_config_dir }}"
+    - "{{ prometheus_config_dir }}/conf.d"
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: This isn't used. Please avoid indirection, I don't want to have to search across files to figure out what's going on.

File: roles/prometheus/tasks/main.yml [hunk 13]
```diff
@@ -1,53 +1,113 @@
...
+- name: Download component binaries
+  get_url:
+    url: "{{item}}"
+    dest: "{{ prometheus_bin_dir }}"
+  with_items:
+    - "https://github.com/prometheus/node_exporter/releases/download/v{{ node_exporter_version }}/{{ node_exporter_release_name }}.tar.gz"
+    - "https://github.com/prometheus/alertmanager/releases/download/v{{ alertmanager_version }}/{{ alertmanager_release_name }}.tar.gz"
+    - "https://github.com/prometheus/pushgateway/releases/download/v{{ pushgateway_version }}/{{ pushgateway_release_name }}.tar.gz"
+
+- name: Unarchive prometheus binary tarball
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: Why is Prometheus handled differently to the other binaries?

File: roles/prometheus/tasks/main.yml [hunk 14]
```diff
@@ -1,53 +1,113 @@
...
+    creates: "{{ prometheus_bin_dir }}/{{item}}"
+  with_items:
+  - "{{ node_exporter_release_name }}"
+  - "{{ alertmanager_release_name }}"
+  - "{{ pushgateway_release_name }}"
+
+- name: Install grafana
+  apt: name=grafana state=latest update_cache=true cache_valid_time=600
+
+- name: Created prometheus service environment file
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: These don't apply when systemd is in use

File: roles/prometheus/tasks/main.yml [hunk 15]
```diff
@@ -1,53 +1,113 @@
...
-  notify: restart prometheus
   copy: src={{item}} dest=/etc/prometheus/{{item}}
   with_items:
     - alert.rules
     - life.rules
-
-- name: Create grafana config
-  notify: restart grafana
-  template: src=grafana.ini dest=/etc/grafana/grafana.ini
+  notify: restart prometheus
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: You only need to reload

**Comment by Conorbro** [comment]
@brian-brazil

File: roles/prometheus/files/life.rules [hunk 16]
```diff
@@ -1,1061 +1,106 @@
...
-init{x="1111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="11111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="1111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="11111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
+groups:
+- name: /etc/prometheus/life.rules
+  rules:
+  - record: grid
+    expr: (((((label_replace(grid, "x", "$1", "x", "^1(.*)$") or init * scalar(round((vector(time())
```
Comment (brian-brazil) [inline roles/prometheus/files/life.rules]: Can you preserve the formatting on this expression?

File: roles/prometheus/files/life.rules [hunk 17]
```diff
@@ -1,1061 +1,106 @@
...
-init{x="1111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="11111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="1111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="11111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="1111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
-init{x="11111111111111111111111111111111",y="11111111111111111111111111111111"} = 1
+groups:
+- name: /etc/prometheus/life.rules
```
Comment (brian-brazil) [inline roles/prometheus/files/life.rules]: Just "life" would be better

File: roles/prometheus/tasks/main.yml [hunk 18]
```diff
@@ -1,53 +1,88 @@
...
+    state=directory
+    path={{ item }}
+    owner=root
+    group=root
+  with_items:
+    - "/opt/prometheus"
+    - "/etc/prometheus"
+    - "/etc/prometheus/rules"
+    - "/etc/prometheus/rules/tgroups"
+    - "/var/lib/prometheus"
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: I don't think you want this owned as root

File: roles/prometheus/tasks/main.yml [hunk 19]
```diff
@@ -1,53 +1,88 @@
...
+  file: >
+    state=directory
+    path={{ item }}
+    owner=root
+    group=root
+  with_items:
+    - "/opt/prometheus"
+    - "/etc/prometheus"
+    - "/etc/prometheus/rules"
+    - "/etc/prometheus/rules/tgroups"
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: This isn't used

File: roles/prometheus/tasks/main.yml [hunk 20]
```diff
@@ -1,53 +1,88 @@
...
+    - "/var/lib/prometheus"
+    - "/var/log/prometheus"
+    - "/etc/prometheus/console_libraries"
+    - "/etc/prometheus/consoles"
+    - "/opt/prometheus/current"
+    - "/etc/grafana"
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: The grafana package will create this

File: roles/prometheus/tasks/main.yml [hunk 21]
```diff
@@ -1,53 +1,88 @@
...
+    owner=root
+    group=root
+  with_items:
+    - "/opt/prometheus"
+    - "/etc/prometheus"
+    - "/etc/prometheus/rules"
+    - "/etc/prometheus/rules/tgroups"
+    - "/var/lib/prometheus"
+    - "/var/log/prometheus"
+    - "/etc/prometheus/console_libraries"
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: Leave these where the tarball creates them

File: roles/prometheus/tasks/main.yml [hunk 22]
```diff
@@ -1,53 +1,88 @@
...
+    group: root
+    mode: 0644
   with_items:
     - prometheus
+    - node_exporter
     - pushgateway
     - alertmanager
-    - node-exporter
-    - grafana
+  when: ansible_service_mgr == 'systemd'
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: You shouldn't need this, we know we're using systemd

File: roles/prometheus/tasks/main.yml [hunk 23]
```diff
@@ -1,53 +1,88 @@
...
 
-- name: Create prometheus config
-  notify: restart prometheus
-  template: src=prometheus.yml dest=/etc/prometheus/prometheus.yml
+- name: Create menu.lib
```
Comment (brian-brazil) [inline roles/prometheus/tasks/main.yml]: This comes from the tarball

File: roles/prometheus/templates/etc/systemd/system/prometheus.service.j2 [hunk 24]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Prometheus
+After=network.target
+
+[Service]
+Type=simple
+EnvironmentFile=/etc/default/prometheus
+PIDFile=/var/run/prometheus.pid
+ExecReload=/bin/kill -HUP $MAINPID
+ExecStart=/opt/prometheus/current/prometheus-2.0.0-rc.1.linux-amd64/prometheus --config.file=/etc/prometheus/prometheus.yml --web.listen-address=:9090 --web.external-url=demo.robustperception.io --web.console.libraries=/opt/prometheus/current/prometheus-2.0.0-rc.1.linux-amd64/console_libraries --web.console.templates=/opt/prometheus/current/prometheus-2.0.0-rc.1.linux-amd64/consoles
```
Comment (brian-brazil) [inline roles/prometheus/templates/etc/systemd/system/prometheus.service.j2]: I'd template out the version numbers here, they appear in a few too many places so it's likely someone would forget to update them. You could also use a symlnk.

Where's the storage directory?

**Comment by Conorbro** [comment]
@brian-brazil

File: roles/base/tasks/main.yml [hunk 25]
```diff
@@ -1,26 +1,39 @@
...
+    system: yes
+    state: present
+
+- name: create prometheus system user
+  user:
+    name: prometheus
+    system: yes
+    shell: "/sbin/nologin"
+    group: prometheus
+    groups: "ssl-cert"
```
Comment (brian-brazil) [inline roles/base/tasks/main.yml]: Why does it need this?

File: roles/base/tasks/main.yml [hunk 26]
```diff
@@ -1,26 +1,39 @@
...
 
 - name: Disable password ssh logins
   notify: restart sshd
   lineinfile: dest=/etc/ssh/sshd_config regexp="^#?PasswordAuthentication .*" line="PasswordAuthentication no"
+
+- name: create prometheus system group
+  group:
+    name: prometheus
+    system: yes
+    state: present
```
Comment (brian-brazil) [inline roles/base/tasks/main.yml]: This is the default, no need to mention it

File: roles/prometheus/templates/etc/systemd/system/alertmanager.service.j2 [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+Type=simple
+PIDFile=/var/run/alertmanager.pid
+ExecReload=/bin/kill -HUP $MAINPID
+ExecStart=/opt/prometheus/current/alertmanager-{{ alertmanager_version }}.linux-amd64/alertmanager -config.file=/opt/prometheus/current/alertmanager-{{ alertmanager_version }}.linux-amd64/simple.yml
+Restart=always
+SyslogIdentifier=alertmanager]
```
Comment (brian-brazil) [inline roles/prometheus/templates/etc/systemd/system/alertmanager.service.j2]: Trailing ]

You're missing the other settings that'd send the logs to syslog which would make this useful

**Comment by Conorbro** [comment]
@brian-brazil

File: roles/prometheus/files/alert.rules [hunk 28]
```diff
@@ -1,2 +1,5 @@
-ALERT ExampleAlertAlwaysFiring
-  IF sum by (job)(up) == 1
+groups:
+- name: /etc/prometheus/alert.rules
+  rules:
+  - alert: ExampleAlertAlwaysFiring
+    expr: sum(up) BY (job) == 1
```
Comment (brian-brazil) [inline roles/prometheus/files/alert.rules]: We always put the by/without before the expression.

File: roles/base/tasks/main.yml [hunk 29]
```diff
@@ -1,26 +1,37 @@
...
-
 - name: Add grafana apt key
   notify: update apt
   apt_key: url=https://packagecloud.io/gpg.key
 
 - name: Disable password ssh logins
   notify: restart sshd
   lineinfile: dest=/etc/ssh/sshd_config regexp="^#?PasswordAuthentication .*" line="PasswordAuthentication no"
+
+- name: create prometheus system group
```
Comment (brian-brazil) [inline roles/base/tasks/main.yml]: The user and group creation should go in the prometheus role

**Comment by conr** [comment]
@brian-brazil

**Pull request merged**